network/tests/topology.rs

```diff
@@ -0,0 +1,301 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use snarkos_network::Node;
+use snarkos_testing::{
+    network::{
+        test_environment,
+        test_node,
+        topology::{connect_nodes, Topology},
+        TestSetup,
+    },
+    wait_until,
+};
+
+const N: usize = 50;
+const MIN_PEERS: u16 = 5;
+const MAX_PEERS: u16 = 10;
+
+async fn test_nodes(n: usize, setup: TestSetup) -> Vec<Node> {
+    let mut nodes = Vec::with_capacity(n);
+
+    for _ in 0..n {
+        let environment = test_environment(setup.clone());
+        let mut node = Node::new(environment).await.unwrap();
+
+        node.establish_address().await.unwrap();
+        nodes.push(node);
+    }
+
+    nodes
+}
+
+async fn start_nodes(nodes: &[Node]) {
+    for node in nodes {
+        // Nodes are started with a slight delay to avoid having peering intervals in phase (this
+        // is the hypothetical real-world worst case scenario).
+        tokio::time::sleep(std::time::Duration::from_millis(10)).await;
+        node.start_services().await;
+    }
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn spawn_nodes_in_a_line() {
+    let setup = TestSetup {
+        consensus_setup: None,
+        peer_sync_interval: 1,
+        ..Default::default()
+    };
+    let mut nodes = test_nodes(N, setup).await;
+    connect_nodes(&mut nodes, Topology::Line).await;
+    start_nodes(&nodes).await;
+
+    // First and Last nodes should have 1 connected peer.
+    wait_until!(
+        5,
+        nodes.first().unwrap().peer_book.read().number_of_connected_peers() == 1
```

Comment:
We should probably move some of these `PeerBook` methods to `Node` in a follow-up soon.